include/ppx/knob.h

```diff
@@ -0,0 +1,284 @@
+    {
+        mCallback = callback;
+        SetValue(defaultValue, true);
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // FlagText format:
+    // --flag_name <true|false> : flag description
+    std::string FlagText() const override;
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <true|false>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    bool GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // If newVal is valid:
+    // - If updateDefault, mDefaultValue is updated
+    // - If newVal is different from mValue:
+    //   - mValue is updated
+    //   - mCallback is triggered
+    void SetValue(bool newVal, bool updateDefault = false);
+
+private:
+    bool mValue;
+    bool mDefaultValue;
+
+    // mCallback will be called on the new bool value every time it is toggled in the UI, or updated with SetValue()
+    std::function<void(bool)> mCallback;
+};
+
+// KnobIntSlider is an int knob that will be displayed as a slider in the UI
+// ImGui sliders can also become input boxes with ctrl + right click
+class KnobIntSlider
+    : public Knob
+{
+public:
+    KnobIntSlider(const std::string& displayName, const std::string& flagName, const std::string& flagDesc, int defaultValue, int minValue, int maxValue, std::function<void(int)> callback = NULL)
+        : Knob(displayName, flagName, flagDesc)
+    {
+        if (minValue >= maxValue) {
+            PPX_LOG_FATAL(flagName << " KnobIntSlider: slider invalid range " << minValue << "~" << maxValue);
+        }
+        mMinValue = minValue;
+        mMaxValue = maxValue;
+        mCallback = callback;
+        SetValue(defaultValue, true);
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // FlagText format:
+    // --flag_name <mMinValue~mMaxValue> : flag description
+    std::string FlagText() const override;
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <int>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    int GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Returns success if newVal is valid, fails otherwise
+    // If newVal is valid:
+    // - If updateDefault, mDefaultValue is updated
+    // - If newVal is different from mValue:
+    //   - mValue is updated
+    //   - mCallback is triggered
+    Result SetValue(int newVal, bool updateDefault = false);
+
+private:
+    int mValue;
+    int mDefaultValue;
+
+    // mValue will be clamped to the mMinValue to mMaxValue range, inclusive
+    int mMinValue;
+    int mMaxValue;
+
+    // mCallback will be called on the new int value every time the slider is deselected in the UI, or updated with SetIntValue()
+    std::function<void(int)> mCallback;
```

Comment:
This could be a `using` statement shared across the whole API. So we have uniform callback patterns (as we do already, modulus the type).
```cpp
template<typename T>
using OnChangeCallback = std::function<void(const T& value)>;
```

Reply:
Let's have the first version without any callback stuff, we still need to determine at what point do we send out updates.

Reply:
The constructor -> SetDefault -> ResetToDefault -> SetValue -> callback call chain is probably unintended.

There's two question:
- Do we want callback during flag/json parsing instead of after it?
- Do we want to callback to trigger after command buffer for the frame has been submitted?